src/csharp/Grpc.Tools.nuspec

```diff
@@ -2,32 +2,70 @@
 <package>
   <metadata>
     <id>Grpc.Tools</id>
```

Comment:
Ignoring changing the location from /tools to /build, are there any limitations in using Grpc.Tools in the "old" way
(e.g. just using a user-provided script and invoking the binaries manually) - I think it is pretty important to make sure that the old workflow still works for users that don't want to change the way of building their project (I think it's fair to require users to update the path from /tools/... to /build/... but not more).
If the "VS integration" turns out not to work well for some users (I think there will always be a scenario...), it is possible to disable it so it doesn't interefere with the manual .proto build process?

Reply:
nit: it's not *just* VS integration, it's a complete integration for VS, standalone MSBuild, .NET.SDK on multiple platforms and mono (I am testing on Windows and Linux).

The short answer is no, just change "tools" to "build".

The long answer is, unfortunately, there is no "old" way to use the tools, without going through hoops and encoding explicit paths to grpc tools. The problem is, there are, again, two different ways of placing NuGet packages, one corresponding to the "old" (non-.NET.SDK projects), when NuGet is a separate tool. The packages are (by default) placed into the `/packages` subdirectory of the solution. This is less of a problem when packaged are libraries, as they are just added with a relative path to the project references. Here's an example from an "old" style project (which we currently only use, with very few exceptions):
```xml
    <Reference Include="Grpc.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d754f35622e28bad, processorArchitecture=MSIL">
      <HintPath>..\..\packages\Grpc.Core.1.7.1\lib\net45\Grpc.Core.dll</HintPath>
    </Reference>
```
The <HintPath> is a misnomer, it is the actual location where the library should be found. Of course, moving projects is a headache, and switching the target version of the framework is a double-headache (note the `\net45\` part, which might or might not work with a different framework). These are created by the NuGet extension inside Visual Studio, and quite brittle.

The "new" build scripts integrate NuGet into the build process. So in a "new" style project, there is a reference to NuGet package ID only, and the build process figures out the correct library folder, if there are multiple. Here's how it looks:
```xml
  <ItemGroup>
    <PackageReference Include="Grpc.Core" Version="1.7.1" />
  </ItemGroup>
```
(happily, VS2017 supports that for the "old" projects, but I digress).

The tools as packaged are hardly useful, *because they are not discoverable* by the build process. First of all, the solution-level `packages` folder is only a default. The user can change the location in his Nuget.config so that, for one example, all packages from all builds are placed under the same directory and shared. This would make the path to the tools totally disconnected from, and absolutely undiscoverable for the build. While I can think of a double-horrific hack of traversing the tree up, looking for a `package` directory, finding one (or more) that matches the `Grpc.Core.*` pattern, then maybe selecting one with the hifghest version (but oh boy, how??), this is brittle by itself and also will break with the common location for projects.

Worse news is that the common location *is* actually the default in .NET.SDK projects. Here's a sample project under Linux that I used while developing these scripts (version 1.9.9.5 is just fake large version to pick the package off my local feed):
```
kkm@yupana:~/work/dotnet/hwapp$ cat hwapp.csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Google.Protobuf" Version="3.4.1" />
    <PackageReference Include="Grpc.Tools" Version="1.9.9.5" />
  </ItemGroup>

</Project>
kkm@yupana:~/work/dotnet/hwapp$ dotnet nuget locals global-packages --list
info : global-packages: /home/kkm/.nuget/packages/
kkm@yupana:~/work/dotnet/hwapp$ ll /home/kkm/.nuget/packages/grpc.tools
total 4.0K
drwxrwxr-x 3 kkm kkm 4.0K Oct 31 17:11 1.9.9.5/
```
So you see where the guy is hiding. NuGet is absolutely not doing any good here; if I wanted to compile some .proto files using a script, I would much prefer to download a .tar.gz file with the tools and other required files, than trying to parse the output of `dotnet nuget locals global-packages --list` in a script (and then the problem of multiple versions does not go away).

All in all, he point of this "long answer" exposition is to show that packaging, quite unfortunately, simply *cannot* get worse than it is. It's currently rather dysfunctional, a quite inconvenient .zip file with extra metadata stuff in it (.nuget is a .zip) that is better unpacked in a known location that pulled automatically from within a project using the tools.

(As a side note, the correct way of the simplest case for tool discovery is including a .props or .targets file that sets a variable based on the predefined variable `$(MSBuildThisFileDirectory)` which is set to the location of the current, be it the root or <Import>'ed, project file. `.props` is for something that user can override in project, and `.targets` is for things we want  more control of).

Reply:
Thanks for the comment, I do agree that the "old" way is hardly sufficient, but I'll try to rephrase my question:
1. let's say someone has Grpc.Tools installed in his project and that project has some .proto files that were previously being build using some kind of script (that the user herself wrote)
2. the .proto files don't necessarily have the right directory structure, they are just somewhere in the project directory because until now it didn't really matter.
3. after upgrading to Grpc.Tools from this PR,  the automatic build action for .proto files present in the project will kick in and likely fail. And maybe the user isn't ready to restructure her project to fix those build issues or there might be some issues that are tricky to solve.

My question is basically what can we do to prevent users' existing projects from breaking once they upgrade the Grpc.Tools package? Is there a simple way to disable to .proto build actions?

Reply:
I amsorry, I seem to be failing to communicate properly how dire the problem actually is. That someone is me, and everyone else out there. That is the whole point: compile *does* necessarily break on a package upgrade, because version. We cannot prevent that breakage, and could not from day one. Here's the Git history of the file imported by other projects for the last 3 upgrades. All changes you see are necessarily manual:

```diff
$ git show -G packages -U1 --minimal cde0e90 5c74249 811223e -- ProtoBuf.targets
commit cde0e90cff406827d16e49c48fc33c448f284327
Author: kkm <kkm@smartaction.com>
Date:   Thu Oct 20 18:44:11 2016 -0700

    Upgrade gRPC package version to 1.0.0

diff --git a/cssrc/ProtoBuf.targets b/cssrc/ProtoBuf.targets
index 7a420a7..351575d 100644
--- a/cssrc/ProtoBuf.targets
+++ b/cssrc/ProtoBuf.targets
@@ -6,4 +6,4 @@
   <PropertyGroup>
-    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.0.15.0\tools\windows_x86\</ProtoCompilerPath>
-    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.0.15.0\tools\windows_x86\</ProtoCsPluginPath>
+    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.1.0.0\tools\windows_x86\</ProtoCompilerPath>
+    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.1.0.0\tools\windows_x86\</ProtoCsPluginPath>
   </PropertyGroup>

commit 5c74249d264c322b7cc5c90d7013f0e4aa446b77
Author: kkm <kkm@smartaction.com>
Date:   Sat Oct 21 20:35:38 2017 -0700

    Update gRPC and protobuf packages for C# projects

diff --git a/cssrc/ProtoBuf.targets b/cssrc/ProtoBuf.targets
index 351575d..84aabc9 100644
--- a/cssrc/ProtoBuf.targets
+++ b/cssrc/ProtoBuf.targets
@@ -6,4 +6,4 @@
   <PropertyGroup>
-    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.1.0.0\tools\windows_x86\</ProtoCompilerPath>
-    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.1.0.0\tools\windows_x86\</ProtoCsPluginPath>
+    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.1.6.1\tools\windows_x86\</ProtoCompilerPath>
+    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.1.6.1\tools\windows_x86\</ProtoCsPluginPath>
   </PropertyGroup>

commit 811223eede7ac4177e0bc64fc45070dd8a87d66b
Author: kkm <kkm@smartaction.com>
Date:   Tue Nov 14 19:57:04 2017 -0800

    Upgrade gRPC packages to 1.7.1 (fixed 2 leaks)

diff --git a/cssrc/ProtoBuf.targets b/cssrc/ProtoBuf.targets
index 354e81d..6c2fbf8 100644
--- a/cssrc/ProtoBuf.targets
+++ b/cssrc/ProtoBuf.targets
@@ -6,4 +6,4 @@
   <PropertyGroup>
-    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.1.6.1\tools\windows_x86\</ProtoCompilerPath>
-    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.1.6.1\tools\windows_x86\</ProtoCsPluginPath>
+    <ProtoCompilerPath Condition=" '$(ProtoCompilerPath)' == '' ">..\..\packages\Grpc.Tools.1.7.1\tools\windows_x86\</ProtoCompilerPath>
+    <ProtoCsPluginPath Condition=" '$(ProtoCsPluginPath)' == '' ">..\..\packages\Grpc.Tools.1.7.1\tools\windows_x86\</ProtoCsPluginPath>
   </PropertyGroup>
```
Except those guys who use .NET.SDK are in a worse position: in these "old" project types we use here, I can more or less *guess* where packages are, relative to the project, and commit this file for other team members to use. Users of NET.SDK are out of luck though: the path is not `../packages` or `../../packages` or `../../../packages` for them, but in addition to having the version in it, starts with like `C:/users/kkm/...` on my machine vs `C:/users/jtattermush/...` on your machine. Cannot even check it in for others to use.

The only workaround I know of is use the `nuget` command (a real command, this is not available during build by default) with a few switches that tell nuget essentially “download this package as if you were curl(1), and then unzip it to this directory, as if you were unzip(1), and please do not add version to the extract path”. Essentially, download and unpack a zip file. That's for `nuget.exe` on Windows and Mono, and `dotnet nuget` does not support these (but curl and unzip are usually available on the platforms where you cannot run nuget.exe). I do not really know how other people work around the problem with the gRPC/Proto tooling packages in real life, but I can feel their pain.

Reply:
To give a short answer, if someone wants for some reason to continue using Grpc.Tools this way, they will have to replace "tools" with "build" once *in addition* to the version that they have to replace with every upgrade anyways. No big deal at all.

Reply:
@kkm000  I see, but wouldn't all the .proto files under the project directory be automatically attempted to be built by the new MSBuild rules without me having a chance to opt-out (except uninstalling Grpc.Tools from my project)? Or is there an explicit rule I need to add to my .csproj file to make the automagic .proto compilation kick in? (if so, then there's no problem). I haven't actually tried playing with the new version of Grpc.Tools package yet, hence my question - I might be missing something.

Reply:
@jtattermusch Yes, absolutely. Inside Visual Studio, changing a .cs file from "Build" to "None" in properties makes VS generate a piece in the project file (equivalent to):
```xml
  <Compile Remove="foo.cs" />
  <None Include="foo.cs" />
```
Same happens with proto files which are included by default but excluded specifically per-file.

Without VS (.NET.SDK alone, e. g. on Linux) there is no need adding the `Remove`d files to `<None>` (adding to the None collection is done so that VS continues to show the file in the project tree), so you would just write, e.g.
```xml
  <ProtoBuf Remove="*_test.proto" />
```
to exclude a specific pattern, or, why not,
```xml
  <ProtoBuf Remove="@(ProtoBuf)" />
  <ProtoBuf Include="*only_one.proto" />
```
I the end, I would prefer the defaults be in step with the Microsoft vision of these projects (as I grok it), but it is naturally controllable on the level of individual files.

Reply:
Great, I think that's very acceptable, thanks for the explanation. - I think the "way to disable" should be part of the documentation along with the way to use.  Does having a src/csharp/Grpc.Tools/README.md sound like a good place for a some initial Grpc.Tools docs?

Reply:
If upgrading an existing package radically changes the default build process, that sounds like it's not great on the compatibility front.

Can I just check I understand properly: you're currently considering adding this to Grpc.Tools. Any reason not to create a new package of Grpc.Tools.MSBuild (or Google.Protobuf.Tools.MSBuild)? Yes, moving the tool locations in Grpc.Tools is already disruptive, but it's less so than requiring each project to add extra lines to disable something. Having the MSBuild task separate allows for separate versioning etc as well.